### ISSUE TITLE
Proposal: restrict the contexts in which `undefined` is allowed

### ISSUE THREAD

**Issue by hvenev** [ziglang/zig#7977]
The goal of this proposal is to make it easier to reason about and check program behavior.

Undefined values are complicated. LLVM has `undef`, `poison`, and `freeze poison`, which have subtly different semantics.

I believe `undefined` should only be usable as a right-hand side of a variable initialization. It should not be possible to use `undefined` as a function argument or to initialize a struct field. I am not sure whether we should permit using `undefined` as an immediate right-hand side of an assignment.

In any case, any use of an uninitialized variable other than setting it (the whole of it, not just a field) is instant UB. This can be checked by adding an `is_defined` flag to each local variable.

I don't think this is too restrictive. For example, a struct field that can be uninitialized would have type `union { value: T, uninit: void }`. This has the advantage that in non-unsafe mode zig will automatically add a flag to track whether the union contains a `value` or an `uninit`:

```
pub const MaybeData = union {
    uninit: void,
    value: [100]u8,
};

pub const AStruct = struct {
    head: u32,
    data: MaybeData,
};

noinline fn x(p: *AStruct) callconv(.C) void {
    p.* = .{
        .head = 0,
        .data = .{ .uninit = .{} },
    };
}

noinline fn y(p: *AStruct) callconv(.C) void {
    p.* = .{
        .head = 0,
        .data = .{ .value = "\x00".* ** 100 },
    };
}

comptime {
    @export(x, .{.name = "x", .linkage = .Strong});
    @export(y, .{.name = "y", .linkage = .Strong});
}
```

The code generated is also quite decent.

**Comment by windows-h8r** [comment]
> In any case, any use of an uninitialized variable other than setting it (the whole of it, not just a field) is instant UB.

In other words, *reading from it* causes instant UB.

**Comment by hvenev** [comment]
> > In any case, any use of an uninitialized variable other than setting it (the whole of it, not just a field) is instant UB.
> 
> In other words, _reading from it_ causes instant UB.

Yes, reading is always instant UB. However, some other uses are also instant UB:

```
const T = struct {
    a: i32,
    b: i32,
};

var x: T = undefined;
x.a = 42; // UB, x is undefined
```

But this is OK:
```
var x: T = undefined;
x = .{a: 42, b: 54};
```

This way we can detect uses of uninitialized variables when checks are enabled.

**Comment by LakeByTheWoods** [comment]
What would this give you over using optionals? With the proposal you wouldn't be able to pass pointers of undefined vars to c APIs.

**Comment by hvenev** [comment]
> What would this give you over using optionals?

More clear semantics of undefined values. Currently LLVM has some strange behaviors. For example, adding undefined values is fine, but dividing by an undefined value is UB.

> With the proposal you wouldn't be able to pass pointers of undefined vars to c APIs.

You can still do uninitialized struct fields using unions:

```
const T = extern union {
    uninit: void,
    value: U,
};

const V = extern struct {
    a: usize,
    b: T,
    c: ?*W,
};

var x: V = .{
    .a = 42,
    .b = .{
        .uninit = .{},
    },
    .c = null,
};

cfunc(&x);
```

**Comment by LakeByTheWoods** [comment]
Interfacing with c APIs that initialise the passed in pointer are a main use case for `undefined` values. We can't just drop support for it and force application developers to write a new `extern union` for every argument that may be uninitialised.
